src/mysql_protocol.erl

```diff
@@ -1220,6 +1214,96 @@ recv_packet(SockModule, Socket, Timeout, ExpectSeqNum, Acc) ->
             {error, Reason}
     end.
 
+-spec send_file(module(), term(), Filename :: binary(), AllowedPaths :: [binary()],
+                SeqNum :: integer()) ->
+    {ok | {error, term()}, NextSeqNum :: integer()}.
+send_file(SockModule, Socket, Filename, AllowedPaths, SeqNum0) ->
+    {Result, SeqNum1} = case is_allowed_path(Filename, AllowedPaths) andalso
+                             file:open(Filename, [read, raw, binary]) of
+        false ->
+            {{error, not_allowed}, SeqNum0};
+        {ok, Handle} ->
+            {ok, SeqNum2} = send_file_chunk(SockModule, Socket, Handle, SeqNum0),
+            ok = file:close(Handle),
+            {ok, SeqNum2};
+        {error, _Reason} = E ->
+            {E, SeqNum0}
+    end,
+    {ok, SeqNum3} = send_packet(SockModule, Socket, <<>>, SeqNum1),
+    {Result, SeqNum3}.
+
+-spec is_allowed_path(file:name_all(), [file:name_all()]) -> boolean().
+is_allowed_path(Path, AllowedPaths) ->
+    Path1 = normalize_split_path(Path),
+    lists:any(
+        fun (AllowedPath) ->
+            AllowedPath1 = normalize_split_path(AllowedPath),
+            is_prefix_path(AllowedPath1, Path1)
+        end,
+        AllowedPaths
+    ).
+
+is_prefix_path([], _Path2) ->
+    true;
+is_prefix_path(_Path1, []) ->
+    false;
+is_prefix_path([Segment1|Path1], Path2=[Segment2|_]) when is_binary(Segment1), is_list(Segment2) ->
+    is_prefix_path([binary_to_list(Segment1)|Path1], Path2);
+is_prefix_path(Path1=[Segment1|_], [Segment2|Path2]) when is_list(Segment1), is_binary(Segment2) ->
+    is_prefix_path(Path1, [binary_to_list(Segment2)|Path2]);
+is_prefix_path([Segment1|Path1], [Segment2|Path2]) ->
+    case string:equal(Segment1, Segment2) of
+        true -> is_prefix_path(Path1, Path2);
+        false -> false
+    end.
+
+-spec normalize_split_path(file:name_all()) -> list().
+normalize_split_path(Path) ->
+    Path1 = filename:flatten(Path),
+    Path2 = case filename:pathtype(Path1) of
+        absolute ->
+            Path1;
+        relative ->
+            {ok, Cwd} = file:get_cwd(),
+            filename:absname_join(Cwd, Path1);
+        volumerelative ->
+            Volume = hd(filename:split(Path1)),
+            {ok, Cwd} = file:get_cwd(Volume),
+            filename:absname_join(Cwd, Path1)
+    end,
+    [Root|RelPath] = filename:split(Path2),
+    normalize_split_path(Root, RelPath, []).
+
+normalize_split_path(Root, [], Acc) ->
+    [Root|lists:reverse(Acc)];
+normalize_split_path(Root, ["."|More], Acc) ->
+    normalize_split_path(Root, More, Acc);
+normalize_split_path(Root, [<<".">>|More], Acc) ->
+    normalize_split_path(Root, More, Acc);
+normalize_split_path(Root, [".."|More], []) ->
+    normalize_split_path(Root, More, []);
+normalize_split_path(Root, [<<"..">>|More], []) ->
```

Comment:
We are already at the top of the path (`[]` in the `Acc` param), can't go any further up. The functions in `file` etc just ignore that, also, so I went with that instead of erroring.

Reply:
Because the root is the parent of itself, for historical reasons. I found this on the Wikipedia article on inode:

*"Historically, it was possible to hard link directories. This made the directory structure into an arbitrary directed graph contrary to a directed acyclic graph. It was even possible for a directory to be its own parent. Modern systems generally prohibit this confusing state, except that the parent of root is still defined as root. The most notable exception to this prohibition is found in Mac OS X (versions 10.5 and higher) which allows hard links of directories to be created by the superuser."*

I wonder what trouble hard links can cause if used as in your example with `/dir2/dir2.1/..` on Mac OS X...

Reply:
No idea... but we had our share with MacOS just recently in `ranch` ^^; Trouble is, I don't have any MacOS at hand, so I only see that it doesn't pass when it gets into the CI... ^^;;;

I'll get back to this PR next week I guess, and I'm now of a mind to just disallow `..` in paths, as you suggest.